### PR TITLE
Add migration tool, guide and add script to test if migration job from api-server to CRD is working

### PR DESCRIPTION
**Pull request by mszostok**
**Description**

Changes proposed in this pr:
- Add migration tool and guide which allow clients to do easily migration from api-server to CRD approach.

    **Scenario**
    - **pre-upgrade hook**
      - Make API Server read-only
      - Check if API Server deployment exists. **If deployment was not found we skip the migration**.
      - Backup Service Catalog custom resources to files in a Persistent Volume.
      - Remove `OwnerReference` fields in all Secrets pointed by any ServiceBinding
      - Remove all Service Catalog resources
    - **upgrade**
      - remove the api-server
      - remove the etcd storage
      - adjust secrets, RBAC etc.
      - upgrade controller manager
      - install webhook server
    - **post-upgrade**
      - Scale down the Controller Manager to avoid any resource processing while backing up resources.
      - Restore all resources. The migration tool sets all necessary fields added in the Service Catalog 0.3.0.
      - Add proper `OwnerReference` to all Secrets pointed by ServiceBindings.
      - Scale up the Controller Manager.

      Full documentation with diagrams is added in this PR: https://github.com/kubernetes-sigs/service-catalog/pull/2710/files#diff-c13a535c132e80742d0b5f424960c632

- Add script to test if migration job from api-server to CRD is working (will be executed as CI pipeline).

  **Scenario:**
  ```
  1. Create k8s cluster (kind)
  2. Install Service Catalog from latest api-server release
  3. Created sample data (sample service broker, service instances etc.)
  4. Upgrade Service Catalog to the version built from sources
  5. Execute tests to check if data from step 3 are still working
  6. Shutdown cluster
  ```
  Pipeline in `test-infra`: https://github.com/kubernetes/test-infra/pull/14464
  Example of this pipeline execution: https://storage.googleapis.com/kyma-prow-logs/logs/post-test-sc-upgrade-job/1176903408213299200/build-log.txt
Fixes:
- https://github.com/kubernetes-sigs/service-catalog/issues/2634
- https://github.com/kubernetes-sigs/service-catalog/issues/2660

### CODE REVIEW

**Comment by jberkhahn** [comment]
I think you actually broke some stuff in pkg/controller...

**Comment by mszostok** [comment]
That’s true, there is sth wrong and I’m still working on it:)

**Comment by mszostok** [comment]
/remove-do-not-merge/work-in-progress

**Comment by mszostok** [comment]
@piotrmiskiewicz @jberkhahn PTAL

**Comment by mszostok** [comment]
Waiting for your reviews :)
/hold

File: docs/migration-apiserver-to-crds.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
+---
+title: Migration from API server to CRDs
+layout: docwithnav
+---
+
+Service Catalog upgrade from version 0.2.x (and earlier) to 0.3.x needs a data migration. 
+This document describes how the migration works and what actions must be performed. 
+
+>**NOTE:**
+Before starting the migration, make sure that you performed a full backup of your cluster.
```
Comment (jberkhahn) [inline docs/migration-apiserver-to-crds.md]: "you have performed"

File: docs/migration-apiserver-to-crds.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
+---
+title: Migration from API server to CRDs
+layout: docwithnav
+---
+
+Service Catalog upgrade from version 0.2.x (and earlier) to 0.3.x needs a data migration.
```
Comment (jberkhahn) [inline docs/migration-apiserver-to-crds.md]: "Upgrading Service Catalog from versions 0.2.x (and earlier) to 0.3.x requires a database migration."

File: docs/migration-apiserver-to-crds.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+- Webhook Server was added to perform data validation/mutation using the admission webhooks mechanism
+
+## Upgrade Service Catalog as a Helm release
+
+The Service Catalog Helm release can be upgraded using the `helm upgrade` command, which runs all necessary actions.
+
+![Service Catalog upgrade steps](images/sc-migration-to-crds-steps.svg)
+
+The upgrade to CRDs consists of the following steps:
+1. Make API Server read-only. Before any backup, we should block any resource changes to be sure the backup makes a snapshot. We need to avoid any changes when the migration tool is backuping resources.
```
Comment (jberkhahn) [inline docs/migration-apiserver-to-crds.md]: "backing up resources"

File: docs/migration-apiserver-to-crds.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+Spec.ClusterServiceClassRef, Spec.ClusterServicePlanRef, Spec.ServiceClassRef, Spec.ServicePlanRef.
+These fields are set during an update operation.
+10. Add proper `OwnerReference` to all Secrets pointed by ServiceBindings.
+11. Scale up the Controller Manager. 
+
+>**NOTE:** In step 6, there is no difference between Service Catalog upgrade using your own etcd or the main Kubernetes etcd.
+
+## Upgrade Service Catalog manually
+
+### Backup and deleting resources
```
Comment (jberkhahn) [inline docs/migration-apiserver-to-crds.md]: "Backup and delete resources"

File: docs/migration-apiserver-to-crds.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+```bash
+make build
+```
+
+If you run the migration tool on OSX and want to get a native binary, add the `PLATFORM` environment variable:
+```bash
+PLATFORM=darwin make build
+```
+
+Resulting executable file can be found in the `bin` subdirectory.
```
Comment (jberkhahn) [inline docs/migration-apiserver-to-crds.md]: "The resulting"

**Comment by mszostok** [comment]
@jberkhahn changes applied:)


there is also PR for adding CI for running migration tests: https://github.com/kubernetes/test-infra/pull/14464. 

It was executed on our infra: https://storage.googleapis.com/kyma-prow-logs/logs/post-test-sc-upgrade-job/1176903408213299200/build-log.txt. You can also execute it locally: 
```
./contrib/hack/ci/run-migration-tests.sh
```

**Pull request merged**